Create React App is broken by lack of default polyfills in Webpack 5

Edit by @jordansexton:

To anyone looking for a very good workaround right now, don't use Create React App. Just use the [react-ui-starter](https://github.com/solana-labs/wallet-adapter/tree/master/packages/starter/react-ui-starter) or [material-ui-starter](https://github.com/solana-labs/wallet-adapter/tree/master/packages/starter/material-ui-starter) projects with Parcel, or the [nextjs-starter](https://github.com/solana-labs/wallet-adapter/tree/master/packages/starter/nextjs-starter) project with Next.js, save yourself a huge hassle, and get much faster builds with SWC instead of Babel.

---

**Describe the bug**
Broken build with CRA v5, unable to resolve modules.

**To Reproduce**
Steps to reproduce the behavior:
1. Initialize a new CRA project. `npx create-react-app my-app-test`
2. Install and use any wallet-adapter packages
3. See error

```
BREAKING CHANGE: The request './adapter' failed to resolve only because it was resolved as fully specified
(probably because the origin is strict EcmaScript Module, e. g. a module with javascript mimetype, a '*.mjs' file, or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
```

**Solution**
Relative imports must include a file extension to follow ESM strict mode requirements introduced in Webpack 5. It's valid typescript to include `.js` extensions as part of the import path 🤯 :

example.ts

```typescript
import { Example } from './example.js';
```
compiles to

example.js
```javascript
import { Example } from './example.js';
```

> https://www.npmjs.com/package/typescript-esm / https://github.com/kristoferbaxter/typescript-esm looks promising
> 
> > When you add the tsc-esm compiler following the conclusion of TypeScript's output, this compiler will remap all generated files to use '.mjs' extensions for locally resolved items and rename every output's extension to '.mjs'.
> 
> Sounds like what we want?

Experimented with this. Unfortunately, it doesn't work out of the box. After installing the package, the `lerna build` script fails on the `packages/starter/example` repository;

```shell
lerna ERR! yarn run build exited 1 in '@solana/wallet-adapter-example'
lerna ERR! yarn run build stdout:
$ yarn clean && next build && next export
$ shx rm -rf .next out
info  - Using webpack 5. Reason: Enabled by default https://nextjs.org/docs/messages/webpack5
info  - Checking validity of types...
info  - Creating an optimized production build...
info  - Using external babel configuration from /Users/cogoo/LocalHost/GitHub/solana_projects/wallet-adapter/packages/starter/example/.babelrc.js
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run build stderr:
Failed to compile.

ModuleNotFoundError: Module not found: Error: Can't resolve '@solana/wallet-adapter-react-ui' in '/Users/cogoo/LocalHost/GitHub/solana_projects/wallet-adapter/packages/starter/example/pages'


> Build error occurred
```

I'll be more comfortable adding the extension manually (prebuild) as that's the accepted convention.

https://www.typescriptlang.org/docs/handbook/esm-node.html

--- a/packages/wallets/ledger/package.json
+++ b/packages/wallets/ledger/package.json
@@ -27,6 +27,7 @@
         "postbuild": "echo '{\"type\":\"commonjs\"}' | npx json > lib/cjs/package.json && echo '{\"type\":\"module\"} ' | npx json > lib/esm/package.json"
     },
     "dependencies": {
+        "buffer": "npm:buffer-browserify@^0.2.5",
         "@ledgerhq/hw-transport": "^6.11.2",
         "@ledgerhq/hw-transport-webhid": "^6.11.2",
         "@solana/wallet-adapter-base": "^0.9.1",

--- a/packages/wallets/ledger/src/adapter.ts
+++ b/packages/wallets/ledger/src/adapter.ts
@@ -1,3 +1,5 @@
+import './polyfills/index';
+
 import type Transport from '@ledgerhq/hw-transport';
 import {
     Adapter,

--- a/packages/wallets/ledger/src/polyfills/Buffer.ts
+++ b/packages/wallets/ledger/src/polyfills/Buffer.ts
@@ -1,0 +1,7 @@
+import { Buffer } from 'buffer';
+
+if (typeof window !== 'undefined' && window.Buffer === undefined) {
+    (window as any).Buffer = Buffer;
+}
+
+export {};

--- a/packages/wallets/ledger/src/polyfills/index.ts
+++ b/packages/wallets/ledger/src/polyfills/index.ts
@@ -1,0 +1,3 @@
+'use strict';
+
+import './Buffer';

--- a/packages/wallets/ledger/src/util.ts
+++ b/packages/wallets/ledger/src/util.ts
@@ -1,3 +1,5 @@
+import './polyfills/index';
+
 import Transport, { StatusCodes, TransportStatusError } from '@ledgerhq/hw-transport';
 import { PublicKey, Transaction } from '@solana/web3.js';
 

--- a/packages/wallets/torus/package.json
+++ b/packages/wallets/torus/package.json
@@ -27,6 +27,7 @@
         "postbuild": "echo '{\"type\":\"commonjs\"}' | npx json > lib/cjs/package.json && echo '{\"type\":\"module\"} ' | npx json > lib/esm/package.json"
     },
     "dependencies": {
+        "assert": "npm:assert@^2.0.0",
         "@solana/wallet-adapter-base": "^0.9.1",
         "@solana/web3.js": "^1.20.0",
         "@toruslabs/solana-embed": "^0.0.9",

--- a/packages/wallets/torus/package.json
+++ b/packages/wallets/torus/package.json
@@ -30,6 +30,7 @@
         "assert": "npm:assert@^2.0.0",
         "@solana/wallet-adapter-base": "^0.9.1",
         "@solana/web3.js": "^1.20.0",
+        "stream": "npm:stream-browserify@^3.0.0",
         "@toruslabs/solana-embed": "^0.0.9",
         "@types/keccak": "^3.0.1",
         "@types/readable-stream": "^2.3.11"

